solution support, using fsac workspaces

enricosada opened the pull request:
**REQUIRE https://github.com/fsharp/FsAutoComplete master who add workspacePeek command to FSAC**

idea is to:

- at start, ask fsac  with `workspacePeek` command what is interesting in that directory
- fsac respond with list of *parsed* sln and list of fsprojs
- based on config, is choosen: a sln or the directory or the user choose from list
- solution explorer will show the *full parsed solution graph*, or the directory fsproj. these fsproj are not yet loaded (but will as lazy loading)

Modes using config `FSharp.workspaceMode`

- `sln`  (**default**) -> prefer sln, if any. Use the sln if only one exists, ask user with list if multiple exists. load the directory if noone
- `directory` -> like before, wild west of fsprojs (no sln needed). but the search is done fsac side
- `ionideSearch`-> previous behaviour, search of fsproj vscode side 

additional configurations:

- `FSharp.workspaceModePeekDeepLevel`: The deep level of directory hierarchy when searching for sln/projects. Default 2 ( so search up to `./src/SomeprojectDir/` )

New commands:

- `F#: Change Workspace or Solution` will show the list to change sln or use the directory

Notes:

- if list is shown (like at start), and is pressed esc, the `ionideSearch` is used. same in case of error of `workspacePeek` command

- Added project loading state to `F# Project Explorer` during `project` command ( not yet loaded ->  loading -> failed -> loaded)

- Right click on failed loaded project in solution explorer -> see error info about loading (same printed to F# panel, but easier to discover)
- Right click on successfully loaded project in solution explorer -> see info

Example ask:

![sln](https://user-images.githubusercontent.com/147243/28883600-4e9cb3ba-77af-11e7-9e41-e9fce93f732f.gif)

NOT IN THIS PR:

- be strict? if a source file is not in the fsac workspace, dont do intellisense (project will be not parsed -> will not work -> squiggles)

baronfel commented:
Is is possible to jump-start this process when `code` is invoked?

Say on the command line I invoke `code path/to/solution.sln`, would it be possible to begin the solution load process with that file?

enricosada commented:
Dunno because like that vscodd open just that solution.
But if you `code /path/to` and mode is `sln` will automatically load it, if is the only sln.

From mobile

--
Enrico Sada
@enricosada
________________________________
From: Chet Husk <notifications@github.com>
Sent: Wednesday, August 2, 2017 6:55:30 PM
To: ionide/ionide-vscode-fsharp
Cc: Enrico Sada; Author
Subject: Re: [ionide/ionide-vscode-fsharp] [WIP] solution support, using fsac workspaces (#480)


Is is possible to jump-start this process when code is invoked?

Say on the command line I invoke code path/to/solution.sln, would it be possible to begin the solution load process with that file?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ionide/ionide-vscode-fsharp/pull/480#issuecomment-319733067>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAI_K5337CQaGudqIrgm_jCBMSFRUJAPks5sUKoCgaJpZM4OrVrS>.

Krzysztof-Cieslak commented:
> under ask mode, user choose what to use

I'd want to use by default current mode, I like it, I believe that `.sln` files should be burned with fire. And while I like the PR I think using `sln` should be `opt-in`

Krzysztof-Cieslak commented:
@baronfel: 

>  Is is possible to jump-start this process when code is invoked? Say on the command line I invoke code path/to/solution.sln, would it be possible to begin the solution load process with that file?

Not possible, invoking `code path/to/solution.sln` will fire Code in "single-file" mode, with just the file opened

enricosada commented:
If is really a need (not for shortcuts), can be done with an env var

```
set IONIDE_SLN=my.sln && code .
```

From mobile

--
Enrico Sada
@enricosada
________________________________
From: Krzysztof Cieślak <notifications@github.com>
Sent: Wednesday, August 2, 2017 8:19:24 PM
To: ionide/ionide-vscode-fsharp
Cc: Enrico Sada; Mention
Subject: Re: [ionide/ionide-vscode-fsharp] [WIP] solution support, using fsac workspaces (#480)


@baronfel<https://github.com/baronfel>:

Is is possible to jump-start this process when code is invoked? Say on the command line I invoke code path/to/solution.sln, would it be possible to begin the solution load process with that file?

Not possible, invoking code path/to/solution.sln will fire Code in "single-file" mode, with just the file opened

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ionide/ionide-vscode-fsharp/pull/480#issuecomment-319755609>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAI_K_HlouNjgsLHW6qNyDQBVcMnXG7Qks5sUL2sgaJpZM4OrVrS>.

enricosada commented:
A bit bumpy while loading alll projects...

![sln_yet](https://user-images.githubusercontent.com/147243/28997412-3ad50f14-7a14-11e7-9ee6-c657ca1e5d52.gif)

enricosada commented:
- cleaned up initialization.
- consolidate old and new mode
- now is not bumpy, and display (`not loaded` -> `loading` -> `loaded|failed`)
- added right click with info about why failure or status (useful later for target fw, etc)

![sln_errors](https://user-images.githubusercontent.com/147243/29008077-d848621c-7b10-11e7-86f8-176a65b432e8.gif)

enricosada commented:
Added cmd to change workspace or solution

![sln_change](https://user-images.githubusercontent.com/147243/29008453-a79393d8-7b17-11e7-9bb1-9ca2773f98d7.gif)

enricosada commented:
@Krzysztof-Cieslak this is done for me. i am satified atm with result and UX. 
Open to feedback about UX and code

default is previous behaviour, so i hope can be published, and meanwhile tested by user changing the user setting, before change the default behaviour.

**REQUIRE https://github.com/fsharp/FsAutoComplete/pull/191 who add workspacePeek command to FSAC, that PR is done and in review atm**

enricosada commented on src/Core/Project.fs:
```diff
@@ -140,16 +160,121 @@ module Project =
+        }
+    
+    let private getWorkspaceForMode mode =
+        let peek () = LanguageService.workspacePeek (vscode.workspace.rootPath) 2 (excluded |> List.ofArray)
```
@Krzysztof-Cieslak atm the deep level is hardcoded to 2 (tradeoff for speed).
level 0 is repo root. 1 are repo dirs. 2 subdirs of dirs

In normal directory structure, like `~/src/My.fsproj` or `~/src/test/integration/MyTests.fsproj` i think is enough deep for a default.

To scan whole directory, there is anyway mode `none`, or open a .fs file as usual.

Krzysztof-Cieslak replied:
Choice seems fine. But maybe it would be worth using DU / Enum instead so we know what means the value we're passing here.

Krzysztof-Cieslak commented:
Another problem is that we're on FSAC's `structuredSignature` branch (which is branch out of `cache` branch) and we need to get it merged to master too... (or at least rebased after fsharp/FsAutoComplete#191 is merged)

Krzysztof-Cieslak commented on release/package.json:
```diff
@@ -606,6 +641,7 @@
     "workspaceContains:**/*.fs",
     "workspaceContains:**/*.fsx",
     "workspaceContains:**/*.fsproj",
+    "workspaceContains:**/*.sln",
```
This will mean that Ionide is activated for C#-only projects, do we (and users) really want it?

Krzysztof-Cieslak commented on src/Components/SolutionExplorer.fs:
```diff
@@ -72,10 +118,13 @@ module SolutionExplorer =
         | ProjectReference (_, name, _) -> name
 
+    let private getRoot () =
+        match getSolution () with
```
`defaultArg` ?

vasily-kirichenko replied:
`Option.defaultValue`

enricosada replied:
👍  going with `defaultArg` because `Option.defaultValue` doesnt exists there seems

Krzysztof-Cieslak commented on src/Components/SolutionExplorer.fs:
```diff
@@ -115,7 +182,13 @@ module SolutionExplorer =
                     | ProjectReference _  -> Some "ionide.projectExplorer.projRef"
                     | Reference _  -> Some "ionide.projectExplorer.reference"
-                    | _ -> None
+                    | Folder _ -> None
```
I'd all context values for every type - it may be useful in the future.

enricosada replied:
ok

enricosada commented:
@Krzysztof-Cieslak  yes seen that.
Ihmo we can wait that your https://github.com/fsharp/FsAutoComplete/pull/186 is merged, and we are back on fsac parity

To just test early this PR, is ok to use my PR's fsac.suave.exe with `let devMode = true` (that's what i do anyway).
Because i didnt change how project are loaded, but just the loading sequence

I'll try to check why your PR tests fails this weekend

vasily-kirichenko commented on src/Components/SolutionExplorer.fs:
```diff
@@ -35,21 +42,55 @@ module SolutionExplorer =
+            | None ->
+                Model.ProjectNotLoaded (projPath, (path.basename projPath))
+            | Some p ->
+                p |> getProjectModelByState
```
Why pipe?

enricosada replied:
yes, i use that too much 😄

vasily-kirichenko commented on src/Components/SolutionExplorer.fs:
```diff
@@ -85,16 +134,34 @@ module SolutionExplorer =
+                        if items |> List.exists isProj then
+                            Some (vscode.TreeItemCollapsibleState.Expanded)
+                        else
+                            Some (vscode.TreeItemCollapsibleState.Collapsed)
```
Why parens here and above?

enricosada replied:
👍

vasily-kirichenko commented on src/Components/SolutionExplorer.fs:
```diff
@@ -146,9 +219,10 @@ module SolutionExplorer =
 
-        Project.projectChanged.event.Invoke(fun proj ->
+        Project.workspaceChanged.event.Invoke(fun _ ->
             emiter.fire (unbox ()) |> unbox)
```
wow. pure magic :)

enricosada replied:
yes, copy pasted it

vasily-kirichenko commented on src/Components/SolutionExplorer.fs:
```diff
@@ -195,4 +269,60 @@ module SolutionExplorer =
+                [ "<b>Status:</b> failed to load"; ""
+                  "<b>Error:</b>"
+                  error ]
+                |> String.concat "<br />"
```
Maybe just put `<br />` at the end of each line?

vasily-kirichenko commented on src/Components/SolutionExplorer.fs:
```diff
@@ -195,4 +269,60 @@ module SolutionExplorer =
+            { new TextDocumentContentProvider with
+                  member this.provideTextDocumentContent (uri: Uri) =
+                      match uri.path with
+                      | "/projects/status" ->
```
Is it always in lower case?

enricosada replied:
yes, that `TextDocumentContentProvider` is something registered and internally used by ionide, so we create both the request urls and the handler

```fsharp
let projectStatusUri projectPath = vscode.Uri.parse(sprintf "fsharp-workspace://authority/projects/status?path=%s" (JS.encodeURIComponent(projectPath)))
```

vasily-kirichenko commented on src/Core/DTO.fs:
```diff
@@ -190,6 +191,41 @@ module DTO =
+        Id: string
+        ConfigurationName: string
+        PlatformName: string
+    }
```
Maybe use a `namespace rec` or `module rec`?

baronfel commented on release/package.json:
```diff
@@ -430,6 +454,17 @@
+        "FSharp.workspaceMode": {
+          "type": "string",
+          "default": "none",
+          "description": "Chooce the workspace mode",
```
Small typo here: `chooce` instead of `choose`

Krzysztof-Cieslak commented:
Due to recent changes to project explorer (lot of great quality of life changes) this will need rebasing, sorry for trouble :(

enricosada commented:
will do, np

enricosada commented:
rebased to https://github.com/fsharp/FsAutoComplete/pull/200

Added more info about project, using new info returned by FSAC `project` response

![image](https://user-images.githubusercontent.com/147243/29981127-d3b4d528-8f4c-11e7-97de-a73512b182f9.png)

enricosada commented:
ok, i need just to tweak a bit the option for configuration now

enricosada commented:
Ok, updated PR description.

pretty satified with this one. default are sane, old behaviour exists too if needed.

Let's wait ionide catch up FSAC `master` `HEAD`

Krzysztof-Cieslak commented:
Requires rebasing again, I guess you can also bump FSAC to include all recent fixes

Krzysztof-Cieslak commented:
I'm bit worried about default settings:
* as I've mentioned before I think old mode should be default one and users needs to opt-in for sln.
* search level 2 seems not enough for some structures I've seen in real world

Krzysztof-Cieslak commented:
What will happen if mode is "directory" and user will run "Change workspace or solution" command?

enricosada commented:
> as I've mentioned before I think old mode should be default one and users needs to opt-in for sln.

the sln is automatically loaded, if there only one.
The reason is: if exists a sln, probably is what the use want to open in that repo.

> search level 2 seems not enough for some structures I've seen in real world

i sampled some proj in github, and seems enough. there is also a config for that, if use want to change.
But full free to use 3, if make more sense ( 2 = seach of proj like inside `./src/Project`, 3 = seach of proj like inside `./src/Group/Project` ) 
Also, if use press esc when shown the list, old behaviour (ionide search side) is used

 
> What will happen if mode is "directory" and user will run "Change workspace or solution" command?

Nothing fancy. Another `peek` request, and the user see the list. If choose something, that is loaded. otherwise nothing change. it's a real choose from list, default matter just for initial loading.

enricosada commented:
@Krzysztof-Cieslak so i rebase and squashed.

The last two commit are an enhancemnt ihmo:

- wait to `parse` and `project` the visible files open at start, until all initial loading of projects is done.
- after all the initial loading is done, reparse visible files.

This help because old behaviour try to guess the fsproj based on source file position. This is not enough lot of times, because may be in a relative path. Or for example a project my depends on others p2p ref, so is not enough.
waiting to load all projects give the best UX ihmo, help avoid the red squigless.
NOTE: waiting load of all project mean all `project` are done, good or bad.
